text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary.
+
+#### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new messages.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+#### Intermediary
+
+An "intermediary" receives a message for the purpose of forwarding it to the
+next receiver, which might be another intermediary or a consumer.
+
+### Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```
+                               CONSUMER
+
+                   . . . . . . Settle
+PRODUCER           .              ^
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |            
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+Two layers of context propagation need to be distinguished for messaging
+workflows:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer MUST attach a creation context to each message. The creation context
+MUST be attached in a way so that it is not changed by intermediaries. A
```

Comment:
That's an interesting proposal, let's discuss it on Thursday. It would require packing both `traceparent` and `tracestate` itself into the `msgCreate` section of `tracestate`.

Reply:
The need to encode `tracestate` at a point in time within a `tracestate` key/value pair does make this a lot less attractive. I had assumed the full SpanContext wouldn't be required for a link, but looking at both the Tracing API spec and the protobuf definition for a Span message, I see the `tracestate` is in fact shown.

It isn't clear to me what use `tracestate` has in  link. I wouldn't expect it is used to identify a span in a back end system, would it? Practically speaking if `tracestate` is omitted in a Span, what functionality might be lost?

Reply:
https://www.w3.org/TR/trace-context/#tracestate-header

`tracestate` is used to encode vendor specific trace identification information (which is not covered by `traceparent`). [According to the spec](https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field) (at least that's my reading), `tracestate` always needs to be passed along if it's present. Which makes sense, as the encoded information might be required by some vendors.

[W3C baggage](https://www.w3.org/TR/baggage/) might be another option for doing what you laid out. However, I'm not convinced about putting both `traceparent` and `tracestate` into the baggage headers.

I will check with W3C folks whether they are fine with updating the W3C AMQP and MQTT spec drafts so they meet our requirements. If that's a possibility, I think that's the best way to go.

Reply:
This discussion can be postponed, as supporting the transport context was added as a future work item.